services/alert/migrate_topic_store.go

```diff
@@ -0,0 +1,160 @@
+package alert
+
+import (
+	"fmt"
+
+	"github.com/influxdata/kapacitor/keyvalue"
+	"github.com/influxdata/kapacitor/services/storage"
+	"github.com/pkg/errors"
+)
+
+const (
+	TopicStoreVersionKey = "topic_store_version"
+	TopicStoreVersion2   = "2"
+)
+
+// MigrateTopicStoreV1V2 - Convert a V1 to a V2 topic store.
+// Also ensures that a topic store has a V2 version number set.
+func (s *Service) MigrateTopicStoreV1V2() error {
+	version, err := s.StorageService.Versions().Get(TopicStoreVersionKey)
+	if err != nil && !errors.Is(err, storage.ErrNoKeyExists) {
+		return fmt.Errorf("cannot determine topic store version: %w", err)
+	}
+	if version == TopicStoreVersion2 {
+		s.diag.Info(fmt.Sprintf("Topic Store is already version %s. Cannot upgrade.", TopicStoreVersion2))
+		return nil
+	}
+
+	topicsDAO, err := NewTopicStateKV(s.StorageService.Store(AlertNameSpace))
+	if err != nil {
+		return fmt.Errorf("cannot create version 1 topic store: %w", err)
+	}
+
+	offset := 0
+	const limit = 100
+
+	topicKeys := make([]string, 0, limit)
+	err = s.StorageService.Store(TopicStatesNameSpace).Update(func(txV2 storage.Tx) error {
+		for {
+			topicStates, err := topicsDAO.List("", offset, limit)
+			if err != nil {
+				return fmt.Errorf("cannot read version 1 topic store: %w", err)
+			}
+			for _, ts := range topicStates {
+				topicKeys = append(topicKeys, ts.Topic)
+				txBucket := txV2.Bucket([]byte(ts.Topic))
+				for id, es := range ts.EventStates {
+					data, err := es.MarshalJSON()
+					if err != nil {
+						return fmt.Errorf("error converting event %q in topic %q to JSON: %w", id, ts.Topic, err)
+					}
+					if err = txBucket.Put(id, data); err != nil {
+						return fmt.Errorf("cannot store event %q in topic %q: %w", id, ts.Topic, err)
+					}
+				}
+			}
+			offset += limit
+			if len(topicStates) != limit {
+				break
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	if err = topicsDAO.DeleteMultiple(topicKeys); err != nil {
+		return err
+	}
+	if err = s.StorageService.Versions().Set(TopicStoreVersionKey, TopicStoreVersion2); err != nil {
+		return fmt.Errorf("cannot set topic store version to %s: %w", TopicStoreVersion2, err)
+	}
+	s.diag.Info("Topic Store updated", keyvalue.T{Key: "version", Value: TopicStoreVersion2})
```

Comment:
If this fails (i.e., the upgrade fails), copy the backup file over the current file.  Think about Windows operations if kapa runs on Windows (copy/delete on open files differences)-- @dgnorton

Reply:
New plan for backup file, @dgnorton 

1. Create backup file if version check shows V1
2. If creation of V2 topic store fails, exit, deleting backup (because transaction rollback will leave BoltDB untouched)
3. If deletion of V1 topic store or update to topic store version number fails, replace BoltDB file with backup
4. Ensure backup is deleted on exit

Reply:
Sounds good :+1: